Minor modification to FMV rules for scope and signatures

AlfieRichardsArm opened the pull request:
Hi all,

While attempting to implement FMV in the GCC front-end some questions were raised that I think are worth clarifying here.

This PR changes the rules to use the default function to determine the signature and scope of the versioned function set.

This clears up some cases such as:

```C
int fn (int c = 1);
int __attribute__((target_version("sve"))) fn (int c = 2);

int bar() { return fn(); }
```
Where there are conflicting signatures and which default should be used is not clear at the moment.

```C
int fn (int c[]);
int __attribute__((target_version("default"))) fn (int c[1]) {
}
int __attribute__((target_version("sve"))) fn (int c[2]) {
}
```
Where if this should be considered a conflicting signature is not clear.

```C
int __attribute__((target_version("default"))) fn (int x) {
    return 1;
}

void bar () {
    int __attribute__((target_version("sve2"))) fn (int);
    fn(1);
}
```
Where the scope of multi-versioned functions differs.

And
```C
// TU 1
#import TU2

int fn (int c = 1);

int bar() { return fn(); }

// TU 2

int __attribute__((target_version("sve"))) fn (int c = 2);
int __attribute__((target_version("sve2"))) fn (int c = 2);

int bar() { return fn(); }
```
Where it is possible calls in different TU's could use different default argument values.

---
name: Pull request
about: Technical issues, document format problems, bugs in scripts or feature proposal.

---

<!-- SPDX-FileCopyrightText: Copyright 2021-2022 Arm Limited and/or its affiliates <open-source-office@arm.com> -->
<!-- SPDX-License-Identifier: Apache-2.0 -->

**Thank you for submitting a pull request!**

If this PR is about a bugfix:

Please use the bugfix label and make sure to go through the checklist below.

If this PR is about a proposal:

We are looking forward to evaluate your proposal, and if possible to
make it part of the Arm C Language Extension (ACLE) specifications.

We would like to encourage you reading through the [contribution
guidelines](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md), in particular the section on [submitting
a proposal](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#proposals-for-new-content).

Please use the proposal label.

As for any pull request, please make sure to go through the below
checklist.

Checklist: (mark with ``X`` those which apply)

* [x] If an issue reporting the bug exists, I have mentioned it in the
      PR (do not bother creating the issue if all you want to do is
      fixing the bug yourself).
* [x] I have added/updated the `SPDX-FileCopyrightText` lines on top
      of any file I have edited. Format is `SPDX-FileCopyrightText:
      Copyright {year} {entity or name} <{contact informations}>`
      (Please update existing copyright lines if applicable. You can
      specify year ranges with hyphen , as in `2017-2019`, and use
      commas to separate gaps, as in `2018-2020, 2022`).
* [x] I have updated the `Copyright` section of the sources of the
      specification I have edited (this will show up in the text
      rendered in the PDF and other output format supported). The
      format is the same described in the previous item.
* [x] I have run the CI scripts (if applicable, as they might be
      tricky to set up on non-*nix machines). The sequence can be
      found in the [contribution
      guidelines](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#continuous-integration). Don't
      worry if you cannot run these scripts on your machine, your
      patch will be automatically checked in the Actions of the pull
      request.
* [x] I have added an item that describes the changes I have
      introduced in this PR in the section **Changes for next
      release** of the section **Change Control**/**Document history**
      of the document. Create **Changes for next release** if it does
      not exist. Notice that changes that are not modifying the
      content and rendering of the specifications (both HTML and PDF)
      do not need to be listed.
* [x] When modifying content and/or its rendering, I have checked the
      correctness of the result in the PDF output (please refer to the
      instructions on [how to build the PDFs
      locally](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#continuous-integration)).
* [x] The variable `draftversion` is set to `true` in the YAML header
      of the sources of the specifications I have modified.
* [ ] Please *DO NOT* add my GitHub profile to the list of contributors
      in the [README](https://github.com/ARM-software/acle/blob/main/README.md#contributors-) page of the project.

AlfieRichardsArm commented on main/acle.md:
```diff
@@ -2682,6 +2683,10 @@ The following attributes trigger the multi version code generation:
 * If only the `default` version exist it should be linked directly.
 * FMV may be disabled in compile time by a compiler flag. In this
   case the `default` version shall be used.
+* Scope for a group of multiversioned functions is the scope of the default
```
I completely agree we want to be able to compile in absence of default declaration. I wrote this to exclude the calling or referring to the dispatched function without the default in scope (rather than the declaration).

My model for this is that the default function _is_ the function, and the extra versions amend that function with different versions for different architecture.
 
This way, if you were to remove all the other version implementations and declarations the code base would still compile.

I think a case like https://godbolt.org/z/qW7hx58bf is emblematic of this point. In this TU there is no indication there is a default version, I think this should require a default version to be declared.

Even in the case you cited (https://github.com/llvm/llvm-project/pull/84405#discussion_r1525583647) they do include the declaration of the default version in all TU's, I believe because it is the idiomatic way to do it?

AlfGalf replied:
I acknowledge that I'm very junior in my career and that I am may be missing significant nuance in this.

I see that argument, but I think I disagree. I think we do need to change something. 

If a multi versioned functions signature is is dominated by the default (ie, that's the canonical signature) than I dont think it makes sense to allow calling that function without that signature in scope.

For example
```C
// TU 1
#import TU2

int fn (int c = 1);

int bar() { return fn(); }

// TU 2

int __attribute__((target_version("sve"))) fn (int c = 2);
int __attribute__((target_version("sve2"))) fn (int c = 2);

int bar() { return fn(); }
```
Would be valid and the calls would have different defaults.

I think the above is quite a good example of why the default declaration determining scope is a good idea.

I also don't think its unreasonable to require the other TU's to have hoisted/shared default declarations. That is the standard for C/C++ functions called in multiple files, and if foo calls bar, and bar calls foo, this would also be required no? It seems idiomatic to require it?

jroelofs replied:
I think I would be okay with a rule that only allows calling a multi-versioned function if you can see the default declaration: we can add diagnostics + fixits that help users do the right thing.

AlfieRichardsArm commented on main/acle.md:
```diff
@@ -2696,10 +2701,12 @@ following:
   * Implicitly, without this attribute,
   * or explicitly providing the `default` in the attribute.
-* All instances of the versions shall share the same function
-  signature and calling convention.
```
I agree the wording should be improved, but I believe there is a need for an adjustment here.
This was specifically something requested by upstream C maintainers in GCC.

The justification is that we likely don't want to require exactly the same signatures, as it
is common for declarations and implementations to differ in signature while still being compatible:

Something like
```C++
int __attribute__((target_version("default"))) fn (int x[]);

int __attribute__((target_version("default"))) fn (int x[5]) {
    return 1;
}
```

Or:
```C++
int __attribute__((target_version("default"))) fn (bool x = true);

int __attribute__((target_version("default"))) fn (bool x) {
    return 1;
}
```
For a more CPP example.

(LLVM is happy with both these currently)

But if this is allowed at some level then we need to define if information for one version should be shared with others, eg a case like:

```C++
int __attribute__((target_version("default"))) fn (bool c = true);
int __attribute__((target_version("sve"))) fn (bool c);

int __attribute__((target_version("default"))) fn (bool c) {
    return 1;
}

int bar() {
    fn();
}
```
(Again allowed by LLVM)
Should this default to true for the SVE version? I would assume so but its not stipulated?

We could disallow it completely, but that seems overly restrictive and counter to C and C++ intuitions.

AlfGalf replied:
Oh wow, okay. That example gives me really bad mojo, and I would advocate we should disallow that if possible, though I wont die on that hill. 
I can live with the default signature dominating, but request we document that in the ACLE.

I would advocate for documenting it in the ACLE. I think the purpose of the ACLE is to document this sort of edge case so we can ensure matched behavior between compilers, and this point came up because my natural inclination when implementing it in GCC was to go another direction. 

This was specifically requested by the GCC C front end maintainer for what I imagine was very much this sort of reason.

AlfieRichardsArm replied:
Ah okay I see, thank you. I have changed the patch to document this rule a little clearer.

As this came up with the GCC front end maintainers I would prefer to clarify this case.

labrinea replied:
The wording seems better. Perhaps it is worth adding the example https://godbolt.org/z/Enn15Y93e without the codegen but explaining that the value 2 will be passed when calling.

AlfieRichardsArm replied:
Just to check, do you mean adding this directly to the ACLE?

AlfieRichardsArm replied:
Added some now :)

AlfieRichardsArm commented on main/acle.md:
```diff
@@ -2682,6 +2683,10 @@ The following attributes trigger the multi version code generation:
   case the `default` version shall be used.
+* Scope for a group of multiversioned functions is the scope of the default
+  version declarations.
+* Dispatiching routines can only be optimized out within the translation unit
```
That makes sense. I'll remove this.

AlfieRichardsArm commented on main/acle.md:
```diff
@@ -2696,10 +2701,12 @@ following:
+  that are combined to a single signature for all versions.
 * All the function versions must be declared at the translation
   unit in which the definition of the default version resides.
+* All function declarations must be at file scope level.
```
Hmm namespaces are a good point (I was looking at this from the C point of view)

I mostly wanted to exclude the case with the versioned function declaration in a function body.
I think it creates the impression that a call in the function will be able to access different versions that outside it?

Like in the example you gave, the resolver calls the `sve2` dispatched version of the function, but it isn't visible to the default version so surely it shouldn't? But what should it do?

AlfGalf replied:
Agreed, I think that works :)
I will update the PR

AlfieRichardsArm replied:
Thank you for the suggestion, I have updated the PR wording to be closer to that.

AlfieRichardsArm commented:
I have updated this PR with hopefully better wording more inline with @labrinea 's suggestions.

labrinea commented on main/acle.md:
```diff
@@ -2699,8 +2702,10 @@ following:
-  signature and calling convention.
+* The default version signature is the definitive signature for calling
+  the multiversioned functions.
+  * All other instances of the versions shall share the same calling 
```
The indentation seems off.

labrinea commented on main/acle.md:
````diff
@@ -2699,11 +2704,41 @@ following:
+void (*f)(void) = f_foo;
+```
+
+Additionally, the below example is valid and 2 is passed as the
````
I think we can merge these two examples into one if you still want to keep the first? There is no versioned `f_foo` in your first example.
```C++
__attribute__((target_version("default"))) void f(int c = 1);
__attribute__((target_version("sve"))) void f(int);

void (*ptr)(int) = f;

void caller() {
  f(); // passes 1
  ptr(2); // passes 2
}
```

AlfieRichardsArm replied:
I think I'll remove the first one. Thank you!

labrinea commented on main/acle.md:
```diff
@@ -2699,11 +2704,41 @@ following:
+
+This case is valid as it is possible to use the
+non-default function as the value for a pointer with the type of
+the default function, as shown below.
```
Perhaps you are mistaken here? We cannot explicitly refer to a specific version. Taking the address of a versioned function and assigning it into a pointer will take the address of the ifunc symbol.

jroelofs replied:
Agreed. That said, I'd like to have syntax for that: https://github.com/llvm/llvm-project/issues/84094

AlfieRichardsArm replied:
This example is intended to clarify what is meant by `Non-default versions shall [..] have a type that is convertible to the type of the default version`. 
I intended to show that a non-default function has "convertible" signature if a function of the non-default signature could be assigned to a function pointer with the default functions signature.

This was a suggestion by Richard Sandiford, that we should use existing function pointer rules to define signature/type compatibility. However, I struggled to formalize it and clearly this example has is not clear for you two, let alone non-experts, so unless you have a way to write this clearer, I will remove it.

labrinea commented on main/acle.md:
```diff
@@ -2699,11 +2704,41 @@ following:
+* The default version signature is the signature for calling
+  the multiversioned functions.
+* Non-default versions shall share the same calling convention
+  of the default version and have a type that is convertable to the type
```
spelling 'convertible'?

AlfieRichardsArm replied:
Thank you!

AlfieRichardsArm commented on main/acle.md:
````diff
@@ -2699,11 +2704,41 @@ following:
+For example, in the case below:
+```C
+__attribute__((target_version("default"))) void fn(void);
+__attribute__((target_version("foo"))) void fn(void); // f_foo below
````
Yup, sounds good.

labrinea commented on main/acle.md:
```diff
@@ -2699,11 +2704,26 @@ following:
+int __attribute__((target_version("default"))) fn (int c = 2);
+int __attribute__((target_version("sve"))) fn (int c = 3);
+
+int baz() { return fn(); }
```
Let's rename baz to something else. I suggest either use naming scheme `g` for caller and `f` for callee. Or just rename baz to `caller`. Otherwise LGTM

AlfieRichardsArm replied:
Awesome, done! Thank you so much for your help with this.

vhscampos commented:
If no objections are raised, let's merge this on the 23 December.

rsandifo-arm commented on main/acle.md:
````diff
@@ -2699,11 +2704,26 @@ following:
+int __attribute__((target_version("sve"))) f (int c = 3);
+
+int g() { return f(); }
+```
````
Just a suggestion, but it might also be worth adding an example to say that:
```
int f (int c = 1);
int __attribute__((target_version("default"))) f (int c = 2);
```
is not valid, because the two statements are taken to be declaring the same entity.

AlfieRichardsArm replied:
Thank you for the suggestions, Ive added this.

labrinea commented on main/acle.md:
````diff
@@ -2699,11 +2704,34 @@ following:
+For example, the below is valid and 2 is passed as the
+value for `c` when calling `f`.
+
+```c
````
NIT: Not sure if it matters, but I think default initialization of arguments is only valid in C++ code, so perhaps the snipped should be marked accordingly?

AlfieRichardsArm replied:
👍 Done

labrinea replied:
I think the markdown annotation for C++ is cpp. You can use build_with_docker.sh to generate locally a pdf version of the document. This way you can double check what the right syntax is.

AlfieRichardsArm replied:
Ah apologies. I have changed this to use `cpp`, checked `build_with_docker.sh`, and rebased onto main.

labrinea commented on main/acle.md:
```diff
@@ -2698,11 +2703,34 @@ following:
+* The default version signature is the signature for calling
+  the multiversioned functions.
+* Non-default versions shall share the same calling convention
+  of the default version and have a type that is convertible to the type
```
"the same calling convention of the default version"

s/of/as

labrinea commented on main/acle.md:
```diff
@@ -2687,6 +2689,9 @@ The following attributes trigger the multi version code generation:
 * FMV may be disabled in compile time by a compiler flag. In this
   case the `default` version shall be used.
+* The scope for calling the versioned function is the scope of the
+  default version.
```
Perhaps rephrase to make it easier for the reader? "A versioned function cannot be called unless the declaration of the default version is visible in the scope of the call site".

AlfieRichardsArm commented:
@labrinea Thank you for the feedback, I addressed them and in doing so moved some stuff around slightly so I hope it makes more sense now.

labrinea commented on main/acle.md:
```diff
@@ -2687,6 +2689,15 @@ The following attributes trigger the multi version code generation:
 * FMV may be disabled in compile time by a compiler flag. In this
   case the `default` version shall be used.
+* A versioned function cannot be called unless the declaration of
+  the default version in visible in the scope of the call site.
```
typo `is` visible

labrinea commented:
May we move the sentence https://github.com/ARM-software/acle/blob/main/main/acle.md?plain=1#L2695 here https://github.com/ARM-software/acle/blob/main/main/acle.md?plain=1#L2677 and rephase it as "Functions are allowed to have the same name when annotated with these attributes." Then in https://github.com/ARM-software/acle/blob/main/main/acle.md?plain=1#L2695 we could say "when applied to a function it becomes one of the versions. Multiple function versions may exist in the same or in different translation units."

Now regarding the order of bullet points I think it could be improved to help the reader. For example the second bullet point refers to the default version a bit suddenly. The third bullet point refers to the table of features which mentions 'default' for the first time in this document.

Also some of thebullet points refer to FMV semantics whereas others specifically refer to the FMV attributes. Perhaps it's worth separating those.

This can be done in a separate commit.

AlfieRichardsArm commented:
I agree the bullets can be made clearer. I will look into making another PR for that work shortly.

I addressed your feedback and while reading through made a few other touches to parts hopefully within scope for this change to make it a little clearer.

labrinea commented on main/acle.md:
```diff
@@ -2687,21 +2691,46 @@ The following attributes trigger the multi version code generation:
+  the default version is visible in the scope of the call site.
+* All function versions must be declared at the same scope level.
+* The default version signature is the signature for calling
+  the multiversioned functions.
```
Under the same bullet: "Therefore a versioned function cannot be called unless the declaration of
  the default version is visible in the scope of the call site." Now we can remove the previous bullet. The reasoning makes more sense if these two points are glued together.

AlfieRichardsArm replied:
That makes sense! Done

philgee-oss commented on main/acle.md:
```diff
@@ -2675,10 +2677,12 @@ The following attributes trigger the multi version code generation:
+* Functions are allowed to have the same name and signature when
+  annotated with these attributes.
 * These attributes can be mixed with each other.
+* `name` is the dependent features from the tables below.
```
Don't use below or above, because they can be ambiguous. Best to cross-reference to the section containing the tables (Mapping?)

philgee-oss commented on main/acle.md:
```diff
@@ -2675,10 +2677,12 @@ The following attributes trigger the multi version code generation:
   be generated without these attributes.
-* `name` is the dependent features from the tables below.
 * The dependent features could be joined by the `+` sign.
 * None of these attributes will enable the corresponding ACLE feature(s)
```
attributes enable

philgee-oss commented on main/acle.md:
```diff
@@ -2687,21 +2691,46 @@ The following attributes trigger the multi version code generation:
 * If only the `default` version exist it should be linked directly.
 * FMV may be disabled in compile time by a compiler flag. In this
```
FMV might

philgee-oss commented on main/acle.md:
```diff
@@ -2687,21 +2691,46 @@ The following attributes trigger the multi version code generation:
 * If only the `default` version exist it should be linked directly.
 * FMV may be disabled in compile time by a compiler flag. In this
   case the `default` version shall be used.
```
case, the

philgee-oss commented on main/acle.md:
```diff
@@ -2687,21 +2691,46 @@ The following attributes trigger the multi version code generation:
-* All the function versions must be declared at the translation
-  unit in which the definition of the default version resides.
+
+For example, the below is valid and 2 is used as the default
```
The following example is valid

philgee-oss commented on main/acle.md:
````diff
@@ -2687,21 +2691,46 @@ The following attributes trigger the multi version code generation:
+int g() { return f(); }
+```
+
+Additionally, the below is not valid as the two statements declare
````
Additionally, the following example is not valid because

AlfieRichardsArm commented:
@philgee-oss Hi, thanks for the feedback, however this has been merged as of 6th Jan.
I can open a new PR and address these if you think its worth it?

vhscampos commented:
> @philgee-oss Hi, thanks for the feedback, however this has been merged as of 6th Jan. I can open a new PR and address these if you think its worth it?

Hi @AlfieRichardsArm , don't worry, I will fix all the issues raised by Phil
